sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/BeamSqlEnv.java

```diff
@@ -73,109 +80,182 @@ public static BeamSqlEnv inMemory(TableProvider... tableProviders) {
+        for (Method method : entry.getValue()) {
+          functionSet.add(new SimpleEntry<>(entry.getKey(), UdfImpl.create(method)));
+        }
+      }
+      return this;
     }
-  }
 
-  public boolean isDdl(String sqlStatement) throws ParseException {
-    try {
-      return planner.parse(sqlStatement) instanceof SqlExecutableStatement;
-    } catch (SqlParseException e) {
-      throw new ParseException("Unable to parse statement", e);
+    public BeamSqlEnvBuilder addSchema(String name, TableProvider tableProvider) {
+      if (schemaMap.containsKey(name)) {
+        throw new RuntimeException("Schema " + name + " is registered twice.");
+      }
+
+      schemaMap.put(name, tableProvider);
+      return this;
     }
-  }
 
-  public void executeDdl(String sqlStatement) throws ParseException {
-    try {
-      SqlExecutableStatement ddl = (SqlExecutableStatement) planner.parse(sqlStatement);
-      ddl.execute(getContext());
-    } catch (SqlParseException e) {
-      throw new ParseException("Unable to parse DDL statement", e);
+    public BeamSqlEnvBuilder setCurrentSchema(String name) {
+      currentSchemaName = name;
+      return this;
     }
-  }
 
-  public CalcitePrepare.Context getContext() {
-    return connection.createPrepareContext();
-  }
+    /** Register a UDF function which can be used in SQL expression. */
+    public BeamSqlEnvBuilder registerUdf(String functionName, Class<?> clazz, String method) {
+      functionSet.add(new SimpleEntry<>(functionName, UdfImpl.create(clazz, method)));
 
-  public Map<String, String> getPipelineOptions() {
-    return connection.getPipelineOptionsMap();
-  }
+      return this;
+    }
 
-  public String explain(String sqlString) throws ParseException {
-    try {
-      return RelOptUtil.toString(planner.convertToBeamRel(sqlString));
-    } catch (ValidationException | RelConversionException | SqlParseException e) {
-      throw new ParseException("Unable to parse statement", e);
+    /** Register a UDF function which can be used in SQL expression. */
+    public BeamSqlEnvBuilder registerUdf(String functionName, Class<? extends BeamSqlUdf> clazz) {
+      return registerUdf(functionName, clazz, BeamSqlUdf.UDF_METHOD);
+    }
+
+    public BeamSqlEnvBuilder registerUdf(String functionName, SerializableFunction sfn) {
+      return registerUdf(functionName, sfn.getClass(), "apply");
+    }
+
+    /**
+     * Register a UDAF function which can be used in GROUP-BY expression. See {@link
+     * org.apache.beam.sdk.transforms.Combine.CombineFn} on how to implement a UDAF.
+     */
+    public BeamSqlEnvBuilder registerUdaf(String functionName, Combine.CombineFn combineFn) {
+      functionSet.add(new SimpleEntry<>(functionName, new UdafImpl(combineFn)));
+      return this;
+    }
+
+    /** Load all UDF/UDAF from {@link UdfUdafProvider}. */
+    public BeamSqlEnvBuilder loadUdfUdafFromProvider() {
+      ServiceLoader.<UdfUdafProvider>load(UdfUdafProvider.class)
```

Comment:
Builder construction is not expected to be expensive, i.e. this should not be called until `build()`. Here it should only set the flag.